cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -214,203 +206,312 @@ public PipelinePlan plan(PipelineSpec spec) {
-        phaseConnections.add(new Connection(conditionName, dag2.getSinks().iterator().next(),
-                                            condition));
-      } else {
-        // scenario here is n1-c1-c2-n2. dag 1 is <c1, c2> and dag 2 is <c2, n2>. We just created condition phase
-        // for c2 and the phase connection should be c2->c2.to.n2
-        phaseConnections.add(new Connection(conditionName, dag2Name, condition));
+      String controlNode = dag.getSinks().iterator().next();
+      if (!phases.containsKey(controlNode)) {
+        PipelinePhase.Builder phaseBuilder = PipelinePhase.builder(supportedPluginTypes);
+        PipelinePhase controlPhase = phaseBuilder
+          .addStage(stageSpecs.get(controlNode))
+          .build();
+        phases.put(controlNode, controlPhase);
       }
-      break;
     }
   }
 
   /**
    * This method is responsible for returning {@link Map} of condition and associated connector name.
    * By default each condition will have associated connector named as conditionname.connector. This connector
-   * will hold the data from the previeous phase. However it is possible that multiple conditions are
+   * will hold the data from the previous phase. However it is possible that multiple conditions are
    * chained together. In this case we do not need to create individual connector for each of the child condition
    * but they should have connector same as parent condition.
    * @param conditionNodes Set of condition nodes in the spec
    * @param childToParent Map contains only conditions as keys. The corresponding value represents its immediate parent
+   * @param conditionInputs Map of condition nodes to corresponding inputs
+   * @param conditionOutputs Map of condition nodes to corresponding outputs
+   * @param actionNodes Set of action nodes
    * @return the resolved connectors for each condition
    */
   private Map<String, String> getConnectorsAssociatedWithConditions(Set<String> conditionNodes,
-                                                                    Map<String, String> childToParent) {
+                                                                    Map<String, String> childToParent,
+                                                                    Map<String, Set<String>> conditionInputs,
+                                                                    Map<String, Set<String>> conditionOutputs,
+                                                                    Set<String> actionNodes) {
     Map<String, String> conditionConnectors = new HashMap<>();
     for (String condition : conditionNodes) {
-      // Put the default connector for the condition
-      conditionConnectors.put(condition, condition + ".connector");
+      Set<String> inputs = conditionInputs.get(condition);
+      Set<String> outputs = conditionOutputs.get(condition);
+
+      if (actionNodes.containsAll(inputs) || actionNodes.containsAll(outputs)) {
```

Comment:
think this still fails if you have something like 'condition1 -> condition2 -> source -> sink'